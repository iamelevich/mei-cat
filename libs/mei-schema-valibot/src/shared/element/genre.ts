import * as v from "valibot";
import { StandardTagSchema } from "../../common";
import { AttrFacsimileSchema } from "../../facsimile/attr/facsimile";
import { AttrAuthorizedSchema } from "../attr/authorized";
import { AttrBiblSchema } from "../attr/bibl";
import { AttrCommonSchema } from "../attr/common";
import { AttrLangSchema } from "../attr/lang";
import { ModelTextPhraseLikeLimitedSchema } from "../model/textPhraseLike.limited";

/**
 * Base schema with attribute, to simplify types for GenreSchema
 */
const GenreBaseSchema = v.object({
	...StandardTagSchema.entries,
	...AttrAuthorizedSchema.entries,
	...AttrBiblSchema.entries,
	...AttrCommonSchema.entries,
	...AttrFacsimileSchema.entries,
	...AttrLangSchema.entries,
});

/**
 * Term or terms that designate a category characterizing a particular style, form, or content.
 * @see https://music-encoding.org/guidelines/v5/elements/genre.html
 */
export const GenreSchema = v.lazy(() =>
	v.intersect([GenreBaseSchema, ModelTextPhraseLikeLimitedSchema]),
);

export type GenreData = v.InferOutput<typeof GenreSchema>;
