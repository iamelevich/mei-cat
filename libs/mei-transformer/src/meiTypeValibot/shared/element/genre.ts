import * as v from "valibot";
import { StandardTagSchema } from "../../common";
import { AttrFacsimileSchema } from "../../facsimile";
import {
	AttrAuthorizedSchema,
	AttrBiblSchema,
	AttrCommonSchema,
	AttrLangSchema,
	ModelTextPhraseLikeLimitedSchema,
} from "..";

/**
 * Base schema with attribute, to simplify types for GenreSchema
 */
const GenreBaseSchema = v.object({
	...StandardTagSchema.entries,
	...AttrAuthorizedSchema.entries,
	...AttrBiblSchema.entries,
	...AttrCommonSchema.entries,
	...AttrFacsimileSchema.entries,
	...AttrLangSchema.entries,
});

/**
 * Term or terms that designate a category characterizing a particular style, form, or content.
 * @see https://music-encoding.org/guidelines/v5/elements/genre.html
 */
export const GenreSchema = v.intersect([
	GenreBaseSchema,
	ModelTextPhraseLikeLimitedSchema,
]);

export type GenreData = v.InferOutput<typeof GenreSchema>;
