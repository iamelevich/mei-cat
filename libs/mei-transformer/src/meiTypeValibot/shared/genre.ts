import * as v from "valibot";
import { AnythingObjectSchema } from "../common";

/**
 * Contains a genre designation.
 * @see https://music-encoding.org/guidelines/v5/elements/genre.html
 */
export const GenreSchema = AnythingObjectSchema; // TODO: implement element

export type GenreData = v.InferOutput<typeof GenreSchema>;
