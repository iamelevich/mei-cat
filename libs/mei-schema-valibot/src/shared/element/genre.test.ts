import { describe, expect, it } from "bun:test";
import * as v from "valibot";
import { GenreSchema } from "./genre";

describe("GenreSchema", () => {
	it("should be defined", () => {
		expect(GenreSchema).toBeDefined();
	});

	it("should parse a genre element with text content", () => {
		const parseResult = v.parse(GenreSchema, {
			"#text": "letter",
			"@xml:id": "genre_N21066",
		});

		expect(parseResult).toMatchSnapshot();
	});

	it("should parse a genre element with different text", () => {
		const parseResult = v.parse(GenreSchema, {
			"#text": "opera",
			"@xml:id": "genre_N21067",
		});

		expect(parseResult).toMatchSnapshot();
	});

	it("should parse a genre element with auth attribute", () => {
		const parseResult = v.parse(GenreSchema, {
			"#text": "symphony",
			"@auth": "LCSH",
			"@xml:id": "genre_N21068",
		});

		expect(parseResult).toMatchSnapshot();
	});

	it("should parse a genre element with auth.uri attribute", () => {
		const parseResult = v.parse(GenreSchema, {
			"#text": "concerto",
			"@auth": "LCSH",
			"@auth.uri": "http://id.loc.gov/authorities/subjects/",
			"@xml:id": "genre_N21069",
		});

		expect(parseResult).toMatchSnapshot();
	});

	it("should parse a minimal genre element", () => {
		const parseResult = v.parse(GenreSchema, {
			"#text": "chamber music",
			"@xml:id": "genre_N2106A",
		});

		expect(parseResult).toMatchSnapshot();
	});
});
