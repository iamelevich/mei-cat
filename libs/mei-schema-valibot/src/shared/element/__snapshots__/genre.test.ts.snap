// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`GenreSchema should parse a genre element with text content 1`] = `
{
  "#text": "letter",
  "@xml:id": "genre_N21066",
}
`;

exports[`GenreSchema should parse a genre element with different text 1`] = `
{
  "#text": "opera",
  "@xml:id": "genre_N21067",
}
`;

exports[`GenreSchema should parse a genre element with auth attribute 1`] = `
{
  "#text": "symphony",
  "@auth": "LCSH",
  "@xml:id": "genre_N21068",
}
`;

exports[`GenreSchema should parse a genre element with auth.uri attribute 1`] = `
{
  "#text": "concerto",
  "@auth": "LCSH",
  "@auth.uri": "http://id.loc.gov/authorities/subjects/",
  "@xml:id": "genre_N21069",
}
`;

exports[`GenreSchema should parse a minimal genre element 1`] = `
{
  "#text": "chamber music",
  "@xml:id": "genre_N2106A",
}
`;
